### ISSUE TITLE
[Feature Request] Note duration randomization

### ISSUE THREAD

**Issue by ryukau** [sjaehn/BSEQuencer#22]
A parameter for note duration randomization is nice to have.

One possible implementation is produce some random number with the range in [0.0, 1.0], then multiply it to duration.

```
amount = getDurationRandomizeAmount(); // In [0.0, 1.0].
random = getRandomNumber();            // In [0.0, 1.0).
duration *= (1 - amount * random);
```

For random number generator, probably `std::minstd_rand` is good enough. Also it's fast. If I remember correctly, SoX uses same algorithm with differenct magic number (the one from Numerical Recipes).

- [Linear congruential generator - Wikipedia](https://en.wikipedia.org/wiki/Linear_congruential_generator)

**Comment by sjaehn** [comment]
Thanks for this request. I also think about randomization of the other pad modifiers (octave, velocity) plus randomization of gate and note.

Also required to implement:
* Larger GUI with space for the new features
* Dials with additional range feature (like in B.Harvestr)

**Comment by a13** [comment]
There are existing plugins for that

**Comment by sjaehn** [comment]
Right. And I guess you think about the x42 midifilters. You can really do a lot with them if you combine them in the right way. Post sequencer.

But I think about pad-individual randomization modifiers. Like pad 1 with randomized duration, pad 2 with randomized gate and octave, and pad 3 no randomization. And you can't get pad-specific randomization with a post sequencer plugin.

Many hardware and software sequencers have already got it build-in. Thus, I'm not inventing sth. new. And right again, there are sequencer plugins with this feature too.

**Comment by ryukau** [comment]
Is it possible that an external plugin changes note duration without adding latency? I'm currently learning MIDI, so I'm a bit curious.

At least in DPF, it seems like a plugin can only read/write MIDI data inside of current cycle. So to shorten the note duration by ratio, latency is inevitable.

**Comment by sjaehn** [comment]
Right, good question. It depends on what the plugin should do.

If you want to make a note longer (augment note value), you don't need to add latency. You only have to *delay* the NOTE_OFF signal. Latency would antagonize the delay.

But if you want to make a note shorter (diminish note value), you need latency indeed. Shortening a note means that you have to move the NOTE_OFF signal backwards in time. And you can only do this by the help of latency. Unless both input NOTE_OFF and output NOTE_OFF accidentally are in the same cycle.

I intensively used latency in B.Schaffl for temporarily shorten note duration (and other MIDI signals).

**Comment by ryukau** [comment]
Thanks for the explanation!

It sounds like DPF is inherited LV2 MIDI handling. As far as I know, VST 3 has same limitation which can't read/write note events outside of current cycle.

**Comment by a13** [comment]
> But I think about pad-individual randomization modifiers. Like pad 1 with randomized duration, pad 2 with randomized gate and octave, and pad 3 no randomization. And you can't get pad-specific randomization with a post sequencer plugin.

It's still possible with x42 midifilters, split, randomize, mix

**Comment by sjaehn** [comment]
> It's still possible with x42 midifilters, split, randomize, mix

... and how do yo make it possible if pad1, (pad2), and pad3 are on the same row (thus, same output note, and maybe same velocity) ? ;-)

**Comment by a13** [comment]
> thus, same output note, and maybe same velocity

I see your point now, but there are a lot of free CCs, which can be used for randomization

**Comment by sjaehn** [comment]
Randomization added in 36398c0165f9cab1b803ba8b643a9f10cc9a4f2f . This feature is not visible by default, YET. Drag or scroll just outside the respective dial in the toolbox. A range arc appears. TODO, I'll make the usage of this feature become a bit more obviously.

Also added: Gate (likelihood), note pitch

**Comment by sjaehn** [comment]
Now also arrows added to make it more obviously in 1f74620738bbe48fffb4b285ef2a47700b4fdd76

**Comment by ryukau** [comment]
Many thanks for the addition. I love the combination of octave and duration randomization.

I think bidirectional range dial used for note, octave and velocity is better stopped at ±0. This makes easier to reset to ±0 with mouse drag.

Otherwise it worked perfect to me.

**Comment by sjaehn** [comment]
There was also a bug in the range dial dragging algorithm. Fixed. Also set lower limit for bidirectional dials to 0. Fixed in 4b3fdc220e2eff54f0f8b973675a3e38a53e697f .
